iox_query/src/exec/gapfill.rs

```diff
@@ -604,4 +614,244 @@ mod test {
+            "| b  | 1970-01-01T00:00:01.100Z |    |",
+            "| b  | 1970-01-01T00:00:01.125Z |    |",
+            "+----+--------------------------+----+",
+        ];
+        assert_batches_eq!(expected, &batches);
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_gapfill_two_output_batches() -> Result<()> {
+        test_helpers::maybe_start_logging();
+        let batch = TestBatch {
+            group_cols: vec![vec![Some("a"), Some("a")]],
+            time_col: vec![Some(1_000), Some(1_100)],
+            agg_cols: vec![vec![Some(10), Some(11)]],
+        };
+        let params = get_params_ms(&batch, 25, 975, 1_125);
+        let tc = TestCase {
+            in_batches: vec![batch],
+            batch_size: 4,
+            params,
+        };
+        let result = tc.run().await;
+        assert!(result.is_err()); // not implemented yet
+
+        // assert_eq!(2, batches.len());
+        // assert_eq!((4, 3), (batches[0].num_rows(), batches[1].num_rows()));
+        // let expected = [
+        //     "+----+--------------------------+----+",
+        //     "| g0 | t                        | a0 |",
+        //     "+----+--------------------------+----+",
+        //     "| a  | 1970-01-01T00:00:00.975Z |    |",
+        //     "| a  | 1970-01-01T00:00:01Z     | 10 |",
+        //     "| a  | 1970-01-01T00:00:01.025Z |    |",
+        //     "| a  | 1970-01-01T00:00:01.050Z |    |",
+        //     "| a  | 1970-01-01T00:00:01.075Z |    |",
+        //     "| a  | 1970-01-01T00:00:01.100Z | 11 |",
+        //     "| a  | 1970-01-01T00:00:01.125Z |    |",
+        //     "+----+--------------------------+----+",
+        // ];
+        // assert_batches_eq!(expected, &batches);
+        Ok(())
+    }
+
+    type ExprVec = Vec<Arc<dyn PhysicalExpr>>;
+
+    struct TestBatch {
+        group_cols: Vec<Vec<Option<&'static str>>>,
+        // Stored as millisecods since intervals use millis,
+        // to let test cases be consistent and easier to read.
+        time_col: Vec<Option<i64>>,
+        agg_cols: Vec<Vec<Option<i32>>>,
+    }
+
+    impl TestBatch {
+        fn num_cols(&self) -> usize {
+            self.group_cols.len() + self.agg_cols.len() + 1
+        }
+
+        fn schema(&self) -> SchemaRef {
+            let mut fields = Vec::with_capacity(self.num_cols());
+            for i in 0..self.group_cols.len() {
+                fields.push(Field::new(
+                    &format!("g{i}"),
+                    DataType::Dictionary(Box::new(DataType::Int32), Box::new(DataType::Utf8)),
+                    true,
+                ));
+            }
+            fields.push(Field::new(
+                "t",
+                DataType::Timestamp(TimeUnit::Nanosecond, None),
+                true,
+            ));
+            for i in 0..self.agg_cols.len() {
+                fields.push(Field::new(&format!("a{i}"), DataType::Int32, true));
+            }
+            Arc::new(Schema::new(fields))
+        }
```

Comment:
I think you might be able to use the `SchemaBuilder` here https://github.com/influxdata/influxdb_iox/blob/67f0eb62b6610a155e9cf54d9ac22e22947b62e0/schema/src/builder.rs#L95-L118

Maybe something like (not tested):

```suggestion
        fn schema(&self) -> SchemaRef {
            let mut builder = SchemaBuilder::new();
            for i in 0..self.group_cols.len() {
              builder = builder.tag(format!("g{i})
            }
            let builder = builder.field("t")
            ... similarly for aggs ...
            builder.build()
        }
```

Reply:
Ah, this is really useful!